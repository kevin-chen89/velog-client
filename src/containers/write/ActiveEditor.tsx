import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { RootState } from '../../modules';
import { WriteMode, clearEditor, prepareEdit } from '../../modules/write';
import EditorPanesContainer from './EditorPanesContainer';
import QuillEditorContainer from './QuillEditorContainer';
import { useLocation } from 'react-router';
import qs from 'qs';
import { useQuery } from '@apollo/react-hooks';
import {
  READ_POST_FOR_EDIT,
  ReadPostForEditResponse,
} from '../../lib/graphql/post';
import { safe } from '../../lib/utils';

export type ActiveEditorProps = {};

const ActiveEditor: React.FC<ActiveEditorProps> = () => {
  const mode = useSelector((state: RootState) => state.write.mode);
  const postId = useSelector((state: RootState) => state.write.postId);

  const dispatch = useDispatch();
  const location = useLocation();

  const { id }: { id?: string } = qs.parse(location.search, {
    ignoreQueryPrefix: true,
  });

  const readPostForEdit = useQuery<ReadPostForEditResponse>(
    READ_POST_FOR_EDIT,
    {
      variables: {
        id,
      },
      skip: !!postId || !id, // skip if id is missing
    },
  );

  useEffect(() => {
    return () => {
      dispatch(clearEditor());
    };
  }, [dispatch]);

  const post = safe(() => readPostForEdit.data!.post);
  useEffect(() => {
    if (!post) return;
    dispatch(
      prepareEdit({
        id: post.id,
        title: post.title,
        body: post.body,
        tags: post.tags,
        description: post.short_description,
        urlSlug: post.url_slug,
        series: post.series,
        isPrivate: post.is_private,
        isMarkdown: post.is_markdown,
        isTemp: post.is_temp,
      }),
    );
  }, [dispatch, post]);

  if (id && !post && !postId) return null;

  return mode === WriteMode.MARKDOWN ? (
    <>
      <EditorPanesContainer />
    </>
  ) : (
    <QuillEditorContainer />
  );
};

export default ActiveEditor;
