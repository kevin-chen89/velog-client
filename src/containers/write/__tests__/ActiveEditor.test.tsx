import * as React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import ActiveEditor, { ActiveEditorProps } from '../ActiveEditor';
import { createStore } from 'redux';
import rootReducer from '../../../modules';
import { Provider } from 'react-redux';

describe('ActiveEditor', () => {
  const setup = (props: Partial<ActiveEditorProps> = {}) => {
    const setupStore = () => {
      const store = createStore(rootReducer);
      return store;
    };
    const store = setupStore();
    const utils = render(
      <Provider store={store}>
        <ActiveEditor {...props} />
      </Provider>,
    );
    return {
      ...utils,
    };
  };
  it('renders properly', () => {
    const utils = setup();
    utils.getByTestId('quill'); // initial mode: WYSIWYG
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  it('converts editor properly', async () => {
    const { getByTestId, getByText, queryByTestId } = setup({});
    const convertButton = getByTestId('mdconvert');
    fireEvent.click(convertButton);
    const confirmButton = getByText('확인');
    fireEvent.click(confirmButton);

    getByTestId('codemirror');
  });
});
