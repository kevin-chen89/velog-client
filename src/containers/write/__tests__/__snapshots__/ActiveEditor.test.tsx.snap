// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveEditor matches snapshot 1`] = `
<div>
  <div
    class="sc-eHgmQL yjeCB"
    data-testid="quill"
  >
    <textarea
      class="sc-htpNat sc-hSdWYo vlAbZ"
      placeholder="제목을 입력하세요"
      style="height: 0px;"
      tabindex="1"
    />
    <div
      class="sc-cvbbAY btkKsm"
    />
    <div
      class="sc-chPdSV drnliS"
    >
      <div
        class="sc-kgoBCf eUtUpj"
        contenteditable="true"
        placeholder="태그를 입력하세요"
        tabindex="2"
      />
      <div
        class="sc-kpOJdX kEzcZB"
      >
        <div
          class="inside"
        >
          쉼표 혹은 엔터를 입력하여 태그를 등록 할 수 있습니다.
          <br />
          등록된 태그를 클릭하면 삭제됩니다.
        </div>
      </div>
    </div>
    <div
      class="sc-bxivhb dbhviG"
      id="toolbar"
    >
      <div
        class="sc-ifAKCX llZOEm"
      >
        <button
          class="ql-header sc-bZQynM fkpewe"
          value="1"
        >
          <div
            class="sc-EHOje gffsLS"
          >
            H
            <span>
              1
            </span>
          </div>
        </button>
        <button
          class="ql-header sc-bZQynM fkpewe"
          value="2"
        >
          <div
            class="sc-EHOje gffsLS"
          >
            H
            <span>
              2
            </span>
          </div>
        </button>
        <button
          class="ql-header sc-bZQynM fkpewe"
          value="3"
        >
          <div
            class="sc-EHOje gffsLS"
          >
            H
            <span>
              3
            </span>
          </div>
        </button>
        <button
          class="ql-header sc-bZQynM fkpewe"
          value="4"
        >
          <div
            class="sc-EHOje gffsLS"
          >
            H
            <span>
              4
            </span>
          </div>
        </button>
      </div>
      <div
        class="sc-gzVnrw bHtWFf"
      />
      <div
        class="sc-ifAKCX llZOEm"
      >
        <button
          class="ql-bold sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M15.6 10.79c.97-.67 1.65-1.77 1.65-2.79 0-2.26-1.75-4-4-4H7v14h7.04c2.09 0 3.71-1.7 3.71-3.79 0-1.52-.86-2.82-2.15-3.42zM10 6.5h3c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5h-3v-3zm3.5 9H10v-3h3.5c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5z"
            />
          </svg>
        </button>
        <button
          class="ql-italic sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M10 4v3h2.21l-3.42 8H6v3h8v-3h-2.21l3.42-8H18V4z"
            />
          </svg>
        </button>
        <button
          class="ql-underline sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 17c3.31 0 6-2.69 6-6V3h-2.5v8c0 1.93-1.57 3.5-3.5 3.5S8.5 12.93 8.5 11V3H6v8c0 3.31 2.69 6 6 6zm-7 2v2h14v-2H5z"
            />
          </svg>
        </button>
        <button
          class="ql-strike sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M10 19h4v-3h-4v3zM5 4v3h5v3h4V7h5V4H5zM3 14h18v-2H3v2z"
            />
          </svg>
        </button>
      </div>
      <div
        class="sc-gzVnrw bHtWFf"
      />
      <div
        class="sc-ifAKCX llZOEm"
      >
        <button
          class="ql-blockquote sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M6 17h3l2-4V7H5v6h3zm8 0h3l2-4V7h-6v6h3z"
            />
          </svg>
        </button>
        <button
          class="ql-link sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
            />
          </svg>
        </button>
        <button
          class="ql-image sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
            />
          </svg>
        </button>
        <button
          class="ql-code-block sc-bZQynM fkpewe"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
            />
          </svg>
        </button>
      </div>
      <div
        class="sc-gzVnrw bHtWFf"
      />
      <div
        class="sc-ifAKCX llZOEm"
      >
        <button
          class="sc-bZQynM fkpewe"
          data-testid="mdconvert"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
          >
            <path
              d="M593.85 452.92H46.15C20.7 452.92 0 432.22 0 406.77V105.23c0-25.45 20.7-46.15 46.15-46.15h547.69c25.45 0 46.15 20.7 46.15 46.15v301.54c.01 25.45-20.69 46.15-46.14 46.15zm-440-92.3v-120l61.54 76.92 61.54-76.92v120h61.54V151.38h-61.54l-61.54 76.92-61.54-76.92H92.31v209.23h61.54zM566.15 256h-61.54V151.38h-61.54V256h-61.54l92.31 107.69L566.15 256z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="sc-jWBwVP boWCkI"
    >
      <div
        class="sc-hMqMXs eaqQgC"
      >
        <div
          tabindex="2"
        />
      </div>
    </div>
    <div
      class="sc-brqgnP hWMpZm"
    >
      <div
        class="sc-dxgOiQ jULWpa"
      >
        <button
          class="sc-ckVGcZ bCLYWM sc-VigVT kMJEjy"
          color="lightGray"
        >
          임시저장
        </button>
        <button
          class="sc-ckVGcZ bCLYWM sc-VigVT eGnsUD"
          color="teal"
        >
          출간하기
        </button>
      </div>
    </div>
  </div>
</div>
`;
